python/pyspark/sql/connect/conversion.py

```
@@ -303,8 +307,11 @@ def convert_variant(value: Any) -> Any:
                     isinstance(value, dict)
                     and all(key in value for key in ["value", "metadata"])
                     and all(isinstance(value[key], bytes) for key in ["value", "metadata"])
+                    and not variants_as_dicts
                 ):
                     return VariantVal(value["value"], value["metadata"])
+                elif isinstance(value, VariantVal) and variants_as_dicts:
```

Isn't there a matrix of inputs we could get?
- `value` is `VariantVal` & `variants_as_dicts` is `False`: not handled?
- `value` is `VariantVal` & `variants_as_dicts` is `True`: handled here, returns `dict`
- `value` is `dict` & `variants_as_dicts` is `False`: handled above, returns `VariantVal`
- `value` is `dict` & `variants_as_dicts` is `True`: not handled?

What do we do for the cases we are not handling?

---

Good Question. For now, we should throw an error in the other cases as I am not aware of any code paths we can use to test them. I specifically set `variants_as_dicts` to false in one particular case which was encountered during `createDataFrame`.

To be more specific:

- `value` is `VariantVal` & `variants_as_dicts` is `False`: This was not handled before (`value` is `VariantVal` was not handled at all) and is still not handled => no regression.
- `value` is `dict` & `variants_as_dicts` is `True`: `variants_as_dicts` is `True` is only possible in one codepath - the one where I have set `variants_as_dicts` to true. Earlier, this would have returned Variants and we would see an error later on anyway (in pa.Table.from_arrays). I don't think this can cause any regressions.